projects/canopy/src/lib/paginator/paginator.component.ts

```diff
@@ -0,0 +1,272 @@
+  /**
+   * The selected page index.
+   */
+  pageIndex: number;
+
+  /**
+   * The selected page size.
+   */
+  pageSize: number;
+
+  /**
+   * The previously selected page index.
+   */
+  previousIndex: number;
+}
+
+/**
+ * New paginator default options.
+ */
+export interface PaginationOptions {
+  /**
+   * The page size.
+   */
+  pageSize?: number;
+
+  /**
+   * The list of page sizes.
+   */
+  pageSizes?: Array<number>;
+
+  /**
+   * Flag to hide the page size selector.
+   */
+  hidePageSize?: boolean;
+
+  /**
+   * Flag to show the first and last page buttons.
+   */
+  showFirstLastButtons?: boolean;
+}
+
+/**
+ * Injection token for setting default paginator options.
+ */
+export const LG_PAGINATOR_DEFAULT_OPTIONS = new InjectionToken<PaginationOptions>(
+  'LG_PAGINATOR_DEFAULT_OPTIONS',
+);
+
+let nextId = 0;
+
+/**
+ * A pagination component used to control the number of items shown in a page context.
+ *
+ * @example
+ * <lg-paginator [length]="tableItems.length" (page)="handleSelectPage($event)"></lg-paginator>
+ */
+@Component({
+  selector: 'lg-paginator',
+  templateUrl: './paginator.component.html',
+  styleUrls: ['./paginator.component.scss'],
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class LgPaginatorComponent {
+  /**
+   * Zero based index for the currently selected page.
+   */
+  @Input()
+  get pageIndex(): number {
+    return this._pageIndex;
+  }
+  set pageIndex(index: number) {
+    this._pageIndex = index;
+    this.setState();
+  }
+  private _pageIndex = 0;
+
+  /**
+   * The number of items shown per page.
+   */
+  @Input()
+  get pageSize(): number {
+    return this._pageSize;
+  }
+  set pageSize(size: number) {
+    this._pageSize = size;
+    this.pageIndex = 0;
+    this.setState();
+  }
+  private _pageSize = 10;
+
+  /**
+   * The total length of items available.
+   */
+  @Input()
+  get length(): number {
+    return this._length;
+  }
+  set length(length: number) {
+    this._length = length;
+    this.pageIndex = 0;
+    this.setState();
+  }
+  private _length: number;
+
+  /**
+   * The list of available page sizes to select.
+   */
+  @Input() pageSizes = [10, 20, 50, 100];
+
+  /**
+   * Flag to determine if the page size selection field should be shown.
+   */
+  @Input() hidePageSize = false;
+
+  /**
+   * Flag to determine if the first and last page controls should be shown.
+   */
+  @Input() showFirstLastButtons = false;
+
+  /**
+   * Output that emits a pagination event when the selected page index changes.
+   */
+  @Output() readonly page = new EventEmitter<PaginationEvent>();
+
+  /**
+   * The id attribute of the component host element.
+   */
+  @HostBinding('id')
+  id = `lg-paginator-${nextId++}`;
+
+  /**
+   * The lg-paginator class name added to the host element.
+   */
+  @HostBinding('class.lg-paginator')
+  class = true;
+
+  currentPageStart = 1;
+  currentPageEnd = 1;
```

Comment:
I wonder if `currentPageStart` and `currentPageEnd` could be Inputs to this component? I could imagine a scenario where you'd want to start off at a particular page, for example if the host application put the page in the URL and wants to make it possible for the user to copy and paste the URL and keep their page position. Or if there was a list of FAQs, or something like that, and the host application needed to deep-link to a specific page.

Reply:
Sure, there is the `pageIndex` input to control the selected page externally. Hopefully you can see how that works in the knobs on the story.

Reply:
Ah I see, so `pageIndex` can set the selected page, and `currentPageStart` and `currentPageEnd` are just internal variables used in the logic?

Reply:
Yes, I should probably encapsulate them but they are for internal use only.